tsfc/driver.py

```diff
@@ -351,66 +341,12 @@ def compile_expression_dual_evaluation(expression, to_element, coordinates, *,
-            assert allclose(qweight, 1.0)
-            assert component == ()
-            qpoints.append(qpoint)
-        point_set = PointSet(qpoints)
-        config = kernel_cfg.copy()
-        config.update(point_set=point_set)
-
-        # Allow interpolation onto QuadratureElements to refer to the quadrature
-        # rule they represent
-        if isinstance(to_element, FIAT.QuadratureElement):
-            assert allclose(asarray(qpoints), asarray(to_element._points))
-            quad_rule = QuadratureRule(point_set, to_element._weights)
-            config["quadrature_rule"] = quad_rule
-
-        expr, = fem.compile_ufl(expression, **config, point_sum=False)
-        shape_indices = tuple(gem.Index() for _ in expr.shape)
-        basis_indices = point_set.indices
-        ir = gem.Indexed(expr, shape_indices)
-    else:
-        # This is general code but is more unrolled than necssary.
-        dual_expressions = []   # one for each functional
-        broadcast_shape = len(expression.ufl_shape) - len(to_element.value_shape())
-        shape_indices = tuple(gem.Index() for _ in expression.ufl_shape[:broadcast_shape])
-        expr_cache = {}         # Sharing of evaluation of the expression at points
-        for dual in to_element.dual_basis():
-            pts = tuple(sorted(dual.get_point_dict().keys()))
-            try:
-                expr, point_set = expr_cache[pts]
-            except KeyError:
-                point_set = PointSet(pts)
-                config = kernel_cfg.copy()
-                config.update(point_set=point_set)
-                expr, = fem.compile_ufl(expression, **config, point_sum=False)
-                expr = gem.partial_indexed(expr, shape_indices)
-                expr_cache[pts] = expr, point_set
-            weights = collections.defaultdict(list)
-            for p in pts:
-                for (w, cmp) in dual.get_point_dict()[p]:
-                    weights[cmp].append(w)
-            qexprs = gem.Zero()
-            for cmp in sorted(weights):
-                qweights = gem.Literal(weights[cmp])
-                qexpr = gem.Indexed(expr, cmp)
-                qexpr = gem.index_sum(gem.Indexed(qweights, point_set.indices)*qexpr,
-                                      point_set.indices)
-                qexprs = gem.Sum(qexprs, qexpr)
-            assert qexprs.shape == ()
-            assert set(qexprs.free_indices) == set(chain(shape_indices, *argument_multiindices))
-            dual_expressions.append(qexprs)
-        basis_indices = (gem.Index(), )
-        ir = gem.Indexed(gem.ListTensor(dual_expressions), basis_indices)
+    # Call dual_evaluation method in FInAT
+    ir_shape = element.dual_evaluation(UFLtoGEMCallback(expression, kernel_cfg, complex_mode))
+    broadcast_shape = len(expression.ufl_shape) - len(element.value_shape)
+    shape_indices = tuple(gem.Index() for _ in expression.ufl_shape[:broadcast_shape])
+    basis_indices = (gem.Index(), )
```

Comment:
I _think_ that the `shape_indices` part (which was a hack when finat didn't provide the dual) should go away and we should `create_element` not `create_base_element` in Firedrake when calling this function.

Reply:
Doesn't Firedrake already use `create_element`? I think only `compile_python_kernel` uses `create_base_element`.

I'm not too familiar with the kernel. What does the kernel do with `shape_indices` and `basis_indices`?

Reply:
See [this firedrake commit](https://github.com/firedrakeproject/firedrake/pull/1743/commits/1b86dbfd039cb7bd11fcf7e4ad6a1fea7a8b0f73).